`NullPointerException: Attempt to invoke virtual method 'boolean android.graphics.Bitmap.compress'`

**Describe the bug**

I am getting this crash in `file_picker`:

```console
03-06 23:31:01.715  1461  1733 W MediaProvider: isAppCloneUserPair for user 0: false
03-06 23:31:01.717  1461  1733 V MediaProvider: Inserted id:1000000047 less than next row id backup:1000001000.
03-06 23:31:01.722  4515  4654 E BitmapFactory: Unable to decode stream: java.lang.NullPointerException
03-06 23:31:01.730  4515  4654 E AndroidRuntime: FATAL EXCEPTION: Thread-5
03-06 23:31:01.730  4515  4654 E AndroidRuntime: Process: app.clicksocial, PID: 4515
03-06 23:31:01.730  4515  4654 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.graphics.Bitmap.compress(android.graphics.Bitmap$CompressFormat, int, java.io.OutputStream)' on a null object reference
03-06 23:31:01.730  4515  4654 E AndroidRuntime: 	at com.mr.flutter.plugin.filepicker.FileUtils.compressImage(FileUtils.java:106)
03-06 23:31:01.730  4515  4654 E AndroidRuntime: 	at com.mr.flutter.plugin.filepicker.FilePickerDelegate$2.run(FilePickerDelegate.java:104)
03-06 23:31:01.730  4515  4654 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:1012)
```

The code is question is the following (the crash happens on `originalBitmap.compress` because `BitmapFactory.decodeFile` returns `null`):

```java
    public static Uri compressImage(Uri originalImageUri, int compressionQuality,Context context) {
        String originalImagePath = getRealPathFromURI(context,originalImageUri);
       Uri compressedUri=null;
       File compressedFile=null;
        try {
             compressedFile=createImageFile();
            Bitmap originalBitmap = BitmapFactory.decodeFile(originalImagePath);
            String file_path = Environment.getExternalStorageDirectory().getAbsolutePath() +
                    "/FilePicker";
            // Compress and save the image
            FileOutputStream fos = new FileOutputStream(compressedFile);
            originalBitmap.compress(Bitmap.CompressFormat.JPEG, compressionQuality, fos);
            fos.flush();
            fos.close();
            compressedUri=Uri.fromFile(compressedFile);
        }catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return compressedUri;
    }
```

I tried adding some debug statements and re-building the library, but I can't seem to build it, because `javax.naming` is not included in Android:

```log
Launching lib/main.dart on sdk gphone x86 64 in debug mode...
/home/luke/.pub-cache/hosted/pub.dev/file_picker-6.2.0/android/src/main/java/com/mr/flutter/plugin/filepicker/FileUtils.java:18: error: package javax.naming does not exist
import javax.naming.Context;
                   ^
/home/luke/.pub-cache/hosted/pub.dev/file_picker-6.2.0/android/src/main/java/com/mr/flutter/plugin/filepicker/FileUtils.java:64: error: cannot find symbol
    public static String getFileName(Uri uri, final Context context) {
                                                    ^
  symbol:   class Context
  location: class FileUtils
```

**Platform**

- [X] Android
- [ ] iOS
- [ ] Web
- [ ] Desktop

**Platform OS version**

API 33

**Flutter Version details**

```log
$ flutter doctor -v
[✓] Flutter (Channel beta, 3.20.0-1.2.pre, on Fedora Linux 39 (Workstation Edition) 6.7.7-200.fc39.x86_64, locale en_US.utf8)
    • Flutter version 3.20.0-1.2.pre on channel beta at /opt/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 1751123cde (2 weeks ago), 2024-02-21 22:06:22 -0800
    • Engine revision 299e852cb9
    • Dart version 3.4.0 (build 3.4.0-99.1.beta)
    • DevTools version 2.31.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /home/luke/Android/Sdk
    • Platform android-34, build-tools 33.0.1
    • Java binary at: /opt/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 17.0.6 (Fedora 17.0.6-2.fc39)
    • cmake version 3.27.7
    • ninja version 1.11.1
    • pkg-config version 1.9.5

[✓] Android Studio (version 2022.2)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 73.0.2
    • Dart plugin version 222.4582
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.87.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.85.20240301

[✓] VS Code (version 1.88.0-insider)
    • VS Code at /usr/share/code-insiders
    • Flutter extension version 3.79.20231130

[✓] Connected device (3 available)
    • sdk gphone x86 64 (mobile) • emulator-5554 • android-x64    • Android 13 (API 33) (emulator)
    • Linux (desktop)            • linux         • linux-x64      • Fedora Linux 39 (Workstation Edition) 6.7.7-200.fc39.x86_64
    • Chrome (web)               • chrome        • web-javascript • Google Chrome 122.0.6261.94

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

--- a/android/src/main/java/com/mr/flutter/plugin/filepicker/FilePickerDelegate.java
+++ b/android/src/main/java/com/mr/flutter/plugin/filepicker/FilePickerDelegate.java
@@ -10,8 +10,8 @@ import android.os.Bundle;
 import android.os.Environment;
 import android.os.Handler;
 import android.os.Looper;
-import android.os.Parcelable;
 import android.os.Message;
+import android.os.Parcelable;
 import android.provider.DocumentsContract;
 import android.util.Log;
 
@@ -19,10 +19,9 @@ import androidx.annotation.VisibleForTesting;
 import androidx.core.app.ActivityCompat;
 
 import java.io.File;
-import java.io.FileNotFoundException;
-import java.io.InputStream;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.Objects;
 
 import io.flutter.plugin.common.EventChannel;
 import io.flutter.plugin.common.MethodChannel;
@@ -77,13 +76,11 @@ public class FilePickerDelegate implements PluginRegistry.ActivityResultListener
 
     @Override
     public boolean onActivityResult(final int requestCode, final int resultCode, final Intent data) {
-
-        if(type == null) {
+        if (type == null) {
             return false;
         }
 
         if (requestCode == REQUEST_CODE && resultCode == Activity.RESULT_OK) {
-
             this.dispatchEventStatus(true);
 
             new Thread(new Runnable() {
@@ -92,16 +89,14 @@ public class FilePickerDelegate implements PluginRegistry.ActivityResultListener
                     if (data != null) {
                         final ArrayList<FileInfo> files = new ArrayList<>();
 
-
                         if (data.getClipData() != null) {
                             final int count = data.getClipData().getItemCount();
                             int currentItem = 0;
                             while (currentItem < count) {
                                  Uri currentUri = data.getClipData().getItemAt(currentItem).getUri();
 
-                                if(type=="image/*" && compressionQuality>0) {
-
-                                    currentUri=FileUtils.compressImage(currentUri,compressionQuality,activity.getApplicationContext());
+                                if (Objects.equals(type, "image/*") && compressionQuality > 0) {
+                                    currentUri = FileUtils.compressImage(currentUri, compressionQuality, activity.getApplicationContext());
                                 }
                                 final FileInfo file = FileUtils.openFileStream(FilePickerDelegate.this.activity, currentUri, loadDataToMemory);
                                 if(file != null) {
@@ -115,8 +110,8 @@ public class FilePickerDelegate implements PluginRegistry.ActivityResultListener
                         } else if (data.getData() != null) {
                             Uri uri = data.getData();
 
-                            if(type=="image/*" && compressionQuality>0) {
-                                uri=FileUtils.compressImage(uri,compressionQuality,activity.getApplicationContext());
+                            if (Objects.equals(type, "image/*") && compressionQuality > 0) {
+                                uri = FileUtils.compressImage(uri, compressionQuality, activity.getApplicationContext());
                             }
 
                             if (type.equals("dir") && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
@@ -230,7 +225,6 @@ public class FilePickerDelegate implements PluginRegistry.ActivityResultListener
         return bundle.getParcelableArrayList("selectedItems");
     }
 
-    @SuppressWarnings("deprecation")
     private void startFileExplorer() {
         final Intent intent;
 
@@ -265,7 +259,7 @@ public class FilePickerDelegate implements PluginRegistry.ActivityResultListener
         }
 
         if (intent.resolveActivity(this.activity.getPackageManager()) != null) {
-            this.activity.startActivityForResult(intent, REQUEST_CODE);
+            this.activity.startActivityForResult(Intent.createChooser(intent, null), REQUEST_CODE);
         } else {
             Log.e(TAG, "Can't find a valid activity to handle the request. Make sure you've a file explorer installed.");
             finishWithError("invalid_format_type", "Can't handle the provided file type.");

--- a/android/src/main/java/com/mr/flutter/plugin/filepicker/FileUtils.java
+++ b/android/src/main/java/com/mr/flutter/plugin/filepicker/FileUtils.java
@@ -94,9 +94,9 @@ public class FileUtils {
 
     public static Uri compressImage(Uri originalImageUri, int compressionQuality, Context context) {
         Uri compressedUri;
-        try {
+        try (InputStream imageStream = context.getContentResolver().openInputStream(originalImageUri)) {
             File compressedFile = createImageFile();
-            Bitmap originalBitmap = BitmapFactory.decodeStream(context.getContentResolver().openInputStream(originalImageUri));
+            Bitmap originalBitmap = BitmapFactory.decodeStream(imageStream);
             // Compress and save the image
             FileOutputStream fos = new FileOutputStream(compressedFile);
             originalBitmap.compress(Bitmap.CompressFormat.JPEG, compressionQuality, fos);

--- a/android/src/main/java/com/mr/flutter/plugin/filepicker/FileUtils.java
+++ b/android/src/main/java/com/mr/flutter/plugin/filepicker/FileUtils.java
@@ -92,28 +92,27 @@ public class FileUtils {
     }
 
 
-    public static Uri compressImage(Uri originalImageUri, int compressionQuality,Context context) {
-        String originalImagePath = getRealPathFromURI(context,originalImageUri);
-       Uri compressedUri=null;
-       File compressedFile=null;
+    public static Uri compressImage(Uri originalImageUri, int compressionQuality, Context context) {
+        Uri compressedUri;
         try {
-             compressedFile=createImageFile();
-            Bitmap originalBitmap = BitmapFactory.decodeFile(originalImagePath);
-            String file_path = Environment.getExternalStorageDirectory().getAbsolutePath() +
-                    "/FilePicker";
+            File compressedFile = createImageFile();
+            Bitmap originalBitmap = BitmapFactory.decodeStream(context.getContentResolver().openInputStream(originalImageUri));
             // Compress and save the image
             FileOutputStream fos = new FileOutputStream(compressedFile);
             originalBitmap.compress(Bitmap.CompressFormat.JPEG, compressionQuality, fos);
             fos.flush();
             fos.close();
-            compressedUri=Uri.fromFile(compressedFile);
-        }catch (FileNotFoundException e) {
+            compressedUri = Uri.fromFile(compressedFile);
+        }
+        catch (FileNotFoundException e) {
             throw new RuntimeException(e);
-        } catch (IOException e) {
+        }
+        catch (IOException e) {
             throw new RuntimeException(e);
         }
         return compressedUri;
     }
+
     private static File createImageFile() throws IOException {
         String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
         String imageFileName = "JPEG_" + timeStamp + "_";

--- a/example/android/app/build.gradle
+++ b/example/android/app/build.gradle
@@ -15,7 +15,7 @@ apply plugin: 'com.android.application'
 apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
 
 android {
-    compileSdkVersion 33
+    compileSdk 33
 
     lintOptions {
         disable 'InvalidPackage'
@@ -24,7 +24,7 @@ android {
     defaultConfig {
         applicationId "com.mr.flutter.plugin.filepicker.example"
         minSdkVersion flutter.minSdkVersion
-        targetSdkVersion 33
+        targetSdk 33
         versionCode 1
         versionName "1.0"
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"

